presto-main/src/main/java/com/facebook/presto/sql/planner/Plan.java

```diff
@@ -13,34 +13,30 @@
  */
 package com.facebook.presto.sql.planner;
 
-import com.facebook.presto.spi.type.Type;
 import com.facebook.presto.sql.planner.plan.PlanNode;
-import com.google.common.collect.ImmutableMap;
-
-import java.util.Map;
 
 import static java.util.Objects.requireNonNull;
 
 public class Plan
 {
     private final PlanNode root;
-    private final Map<Symbol, Type> types;
+    private final TypeProvider types;
 
-    public Plan(PlanNode root, Map<Symbol, Type> types)
+    public Plan(PlanNode root, TypeProvider types)
     {
         requireNonNull(root, "root is null");
         requireNonNull(types, "types is null");
 
         this.root = root;
-        this.types = ImmutableMap.copyOf(types);
+        this.types = types;
```

Comment:
nit: You can also put `requireNonNull(types, "types is null")` on the assignment line (same for `root`)